components/Settings.js

```diff
@@ -391,33 +464,74 @@ class Settings extends React.PureComponent {
 
   handleChange = (key, value) => {
     this.props.onChange(key, value)
-    this.setState({ selectedPreset: null, previousSettings: null })
+    this.props.selectPreset(null)
+    this.setState({ previousSettings: null })
   }
 
-  applyPreset = (index, settings) => {
-    // Do not store functions in previous state
-    const previousSettings = omit(this.props, [
+  getSettings = () =>
+    omit(this.props, [
       'onChange',
       'resetDefaultSettings',
       'applyPreset',
       'removePreset',
-      'format'
+      'format',
+      'getCarbonImage',
+      'selectPreset',
+      'selectedPreset'
     ])
 
-    this.props.applyPreset(omit(settings, ['custom']))
-    this.setState({ selectedPreset: index, previousSettings })
+  applyPreset = (index, preset) => {
+    const previousSettings = this.getSettings()
+
+    this.props.applyPreset(index, preset)
+    this.setState({ previousSettings })
   }
 
   undoPreset = () => {
-    this.props.applyPreset(this.state.previousSettings)
-    this.setState({ selectedPreset: null, previousSettings: null })
+    this.props.applyPreset(null, this.state.previousSettings)
+    this.setState({ previousSettings: null })
   }
 
-  removePreset = index =>
-    this.setState(({ presets }) => ({ presets: presets.filter((_, i) => i !== index) }))
+  removePreset = index => {
+    this.setState(
+      ({ presets }) => ({ presets: presets.filter((_, i) => i !== index) }),
+      () => {
+        const { selectedPreset, selectPreset } = this.props
+        selectPreset(
+          index === selectedPreset ? null : selectedPreset > 0 ? selectedPreset - 1 : selectedPreset
+        )
+        this.savePresets()
+      }
+    )
+  }
+
+  createPreset = async () => {
+    const preset = this.getSettings()
+
+    preset.custom = true
+
+    preset.icon = await this.props.getCarbonImage({ format: 'png', squared: true })
+
+    this.setState(
+      ({ presets }) => ({
+        presets: [preset, ...presets]
+      }),
+      () => {
+        this.props.selectPreset(0)
+        this.savePresets()
+      }
+    )
+  }
+
+  savePresets = () =>
+    savePresets(
+      localStorage,
+      this.state.presets.slice(0, this.state.presets.length - DEFAULT_PRESETS.length)
```

Comment:
Another spot where using index's comes up. This could be presets.filter(preset => preset.custom) which is probably more understandable?

Reply:
Yes please make this change 👍